golang/cosmos/vm/server.go

```diff
@@ -1,23 +1,93 @@
 package vm
 
 import (
+	"context"
 	"fmt"
+	"sync"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
 )
 
-type AgdServer struct{}
+const VmPortMappingContextKey = sdk.ContextKey("vm-port-mapping")
+
+type AgdServer struct {
+	currentCtx    context.Context
+	mtx           sync.Mutex
+	lastPort      int
+	portToHandler map[int]PortHandler
+	portToName    map[int]string
+	nameToPort    map[string]int
```

Comment:
Document the invariants - subject to how you solve the re-registration issue mentioned below. One option is:

```
// portToHandler[i] is nonzero iff portToName[i] is nonzero
// portToName[nameToPort[i]] == i && nameToPort[portToName[s]] == s for all i, s.
```